File: FactoryMethod.cpp
#include <iostream>
using namespace std;

class Button
{
public:
    void render;
    void onClick;
};

class WindowsButton : public Button
{
public:
    void render()
    {
        cout << "WindowsButton is rendered\n";
    }
    void onClick()
    {
        cout << "WindowsButton is clicked\n";
    }
};

class LinuxButton : public Button
{
public:
    void render()
    {
        cout << "LinuxButton is rendered\n";
    }
    void onClick()
    {
        cout << "WindowsButton is clicked\n";
    }
};

class WebButton : public Button
{
public:
    void render()
    {
        cout << "WebButton is rendered\n";
    }
    void onClick()
    {
        cout << "WindowsButton is clicked\n";
    }
};

class Dialog
{
public:
    void build;
    Button render()
    {
        Button okButton = build();
        okButton.onClick();
        okButton.render();
    }
};

class WindowsButton : public Dialog
{
public:
    Button build()
    {
        return WindowsButton();
    }
};

class LinuxButton : public Dialog
{
public:
    Button build()
    {
        return LinuxButton();
    }
};

class WebButton : public Dialog
{
public:
    Button build()
    {
        return WebButton();
    }
};

int main()
{
    return 0;
}

File: factory_method.dart
abstract class Button {
  void render();
  void onClick(Function pressed);
}

class WindowsButton implements Button {
  @override
  void render() {
    print('It\'s WindowsButton');
  }

  @override
  void onClick(Function pressed) {
    pressed.call();
  }
}

class WebButton implements Button {
  @override
  void render() {
    print('It\'s WebButton');
  }

  @override
  void onClick(Function pressed) {
    pressed.call();
  }
}

abstract class Dialog {
  void render() {
    Button okButton = createButton();
    okButton.onClick(() {
      print('Sk');
    });
    okButton.render();
  }

  Button createButton();
}

class WindowsDialog extends Dialog {
  @override
  Button createButton() {
    return WindowsButton();
  }
}

class WebDialog extends Dialog {
  @override
  Button createButton() {
    return WebButton();
  }
}

void main(List<String> args) {
  Dialog newDialog = WebDialog();
  newDialog.createButton().render();
}
